### PR TITLE
Cache eviction policy

### PR DESCRIPTION
**Pull request by luis-gasparschroeder**
### Description: Abstract Eviction Policy Framework

This pull request introduces `EvictionPolicy`, an abstract base class that provides a standardized framework for managing cache eviction. This design ensures that cache size is managed efficiently without blocking critical operations.

#### Design Decisions:

*   **Proactive, Watermark-based Triggering**: Eviction is initiated when the cache size surpasses a configurable high-watermark (e.g., 95% of `max_size`), preventing the cache from becoming completely full.

*   **Efficient Batch Eviction**: Instead of removing items one-by-one, the policy evicts a batch of items based on a configurable percentage (`eviction_percentage`), which is more performant.

*   **Asynchronous & Non-Blocking**: The core eviction logic is executed in a background thread via a `ThreadPoolExecutor`. This ensures that the main application thread remains responsive and is not blocked by the potentially time-consuming removal process.

*   **Thread-Safe**: A `threading.Lock` guarantees that only one eviction process can run at a time. While the lock is active, the cache is in a transient state, and incoming requests may be treated as cache misses. However, this state is extremely brief (typically under 0.005 seconds), so the performance impact is negligible.

*   **Extensible**: Concrete eviction strategies (like LRU, LFU, etc.) can be implemented by subclassing `EvictionPolicy` and providing logic for the `select_victims()` and `update_eviction_metadata()` methods.

data_flow

### CODE REVIEW

File: vcache/main.py [hunk 1]
```diff
@@ -64,7 +64,24 @@ def infer_with_cache_info(
         if system_prompt is None:
             system_prompt = self.vcache_config.system_prompt
 
-        return self.vcache_policy.process_request(prompt, system_prompt)
+        if self.vcache_config.eviction_policy.is_evicting():
```
Comment (luis-gasparschroeder) [inline vcache/main.py:67]: This eviction logic is implemented here to be decoupled from the decision policy logic

File: tests/integration/test_eviction_policy.py [hunk 2]
```diff
@@ -0,0 +1,185 @@
+import time
```
Comment (AlexCuadron) [inline tests/integration/test_eviction_policy.py:1]: The current tests don't cover:

Threading/concurrency issues
Exception handling scenarios
Resource cleanup
Watermark edge cases
Integration failure scenarios

File: vcache/vcache_core/cache/eviction_policy/eviction_policy.py [hunk 3]
```diff
@@ -1,21 +1,183 @@
...
+        """
+        Asynchronously triggers the eviction process if not already running.
+
+        This method acquires a lock to set the evicting state immediately,
+        then submits the eviction task to a background thread.
+
+        Args:
+            cache: The cache instance to perform eviction on.
+        """
+        if self.is_evicting_lock.acquire(blocking=False):
```
Comment (AlexCuadron) [inline vcache/vcache_core/cache/eviction_policy/eviction_policy.py:140]: The metadata is fetched and victims are computed BEFORE submitting to the thread pool, but the lock is released when the method exits (implicitly). This creates a race condition where:

Thread A acquires lock, computes victims based on current cache state
Thread A submits eviction task and exits (lock released)
Thread B can now add new items to cache
Thread A's eviction task runs and removes victims based on stale metadata

Comment (AlexCuadron) [reply vcache/vcache_core/cache/eviction_policy/eviction_policy.py:140]: Also, If cache.get_all_embedding_metadata_objects() or self.select_victims() throws an exception, the lock is never released because these operations happen before the try/finally block in _evict_victims

File: vcache/vcache_core/cache/eviction_policy/eviction_policy.py [hunk 4]
```diff
@@ -1,21 +1,183 @@
...
+            True if the current cache size exceeds the watermark threshold,
+            False otherwise.
+        """
+        if self.is_evicting():
+            return False
+
+        number_of_embeddings_in_cache = cache.vector_db_size()
+        watermark_threshold = self.max_size * self.watermark
+
+        return number_of_embeddings_in_cache > watermark_threshold
```
Comment (AlexCuadron) [inline vcache/vcache_core/cache/eviction_policy/eviction_policy.py:92]: Uses > instead of >=. If watermark is 0.8 and max_size is 10, threshold is 8.0. With 8 items, it won't trigger eviction, but with 9 items it will. This means the cache can grow to max_size before eviction triggers, defeating the watermark's purpose.

File: vcache/vcache_core/cache/eviction_policy/strategies/no_eviction.py [hunk 5]
```diff
@@ -1,22 +1,53 @@
...
+
+    def update_eviction_metadata(self, metadata: EmbeddingMetadataObj) -> None:
+        """Does nothing, as no metadata is needed for this policy."""
+        pass
+
+    def select_victims(self, all_metadata: List[EmbeddingMetadataObj]) -> List[int]:
+        """Never selects any victims, always returns an empty list."""
+        return []
+
+    def evict(self, cache: "Cache", victims: List[int]) -> None:
```
Comment (AlexCuadron) [inline vcache/vcache_core/cache/eviction_policy/strategies/no_eviction.py:42]: Base class evict() method takes only cache, but NoEvictionPolicy.evict() takes cache and victims.

File: vcache/main.py [hunk 6]
```diff
@@ -64,7 +64,24 @@ def infer_with_cache_info(
...
+            response = self.__generate_response(prompt, system_prompt)
+            return False, response, response
+
+        is_cache_hit, response, nn_metadata = self.vcache_policy.process_request(
+            prompt, system_prompt
+        )
+
+        self.vcache_config.eviction_policy.update_eviction_metadata(nn_metadata)
```
Comment (AlexCuadron) [inline vcache/main.py:75]: nn_metadata could potentially be None depending on the policy implementation, but there's no null check.

File: vcache/vcache_core/cache/eviction_policy/strategies/lru.py [hunk 7]
```diff
@@ -1,22 +1,62 @@
...
+        and are prioritized for eviction.
 
         Args:
-            embeddings: The embeddings to rank.
+            all_metadata: A list of all metadata objects in the cache.
+
+        Returns:
+            A list of embedding_ids for the items to be evicted.
+        """
+        sorted_metadata: List[EmbeddingMetadataObj] = sorted(
```
Comment (AlexCuadron) [inline vcache/vcache_core/cache/eviction_policy/strategies/lru.py:38]: Sorting the entire list when only need the N smallest items. For large caches, this is O(n log n) instead of O(n log k) where k is the number to evict.

File: vcache/vcache_core/cache/eviction_policy/strategies/lru.py [hunk 8]
```diff
@@ -1,22 +1,62 @@
...
+            all_metadata,
+            key=lambda meta: meta.last_accessed
+            if meta.last_accessed is not None
+            else datetime.min,
```
Comment (AlexCuadron) [inline vcache/vcache_core/cache/eviction_policy/strategies/lru.py:42]: datetime.min incompatible with timezone-aware datetimes (same for mru)

File: vcache/vcache_core/cache/eviction_policy/strategies/scu.py [hunk 9]
```diff
@@ -0,0 +1,134 @@
...
+        )
+        if max_n_obs == 0:
+            max_n_obs = 1
+
+        utilities: List[Tuple[int, float]] = []
+        for meta in all_metadata:
+            if meta.t_prime is None:
+                distance: float = float("inf")
+            else:
+                n_obs_norm: float = len(meta.observations) / max_n_obs
```
Comment (AlexCuadron) [inline vcache/vcache_core/cache/eviction_policy/strategies/scu.py:118]: Potential division by zero in normalization calculation

File: vcache/vcache_core/cache/cache.py [hunk 10]
```diff
@@ -27,9 +27,9 @@ def __init__(
...
             eviction_policy: Policy for removing items when cache is full.
         """
-        self.embedding_store = embedding_store
-        self.embedding_engine = embedding_engine
-        self.eviction_policy = eviction_policy
+        self.embedding_store: EmbeddingStore = embedding_store
+        self.embedding_engine: EmbeddingEngine = embedding_engine
+        self.eviction_policy: EvictionPolicy = eviction_policy
 
     def add(self, prompt: str, response: str) -> int:
```
Comment (AlexCuadron) [inline vcache/vcache_core/cache/cache.py:34]: Cache.add() doesn't trigger eviction checking

File: vcache/vcache_core/cache/eviction_policy/eviction_policy.py [hunk 11]
```diff
@@ -1,21 +1,183 @@
...
+from typing import TYPE_CHECKING, List, Tuple
+
+if TYPE_CHECKING:
+    from vcache.vcache_core.cache.cache import Cache
+from vcache.vcache_core.cache.embedding_store.embedding_metadata_storage.embedding_metadata_obj import (
+    EmbeddingMetadataObj,
+)
 
 
 class EvictionPolicy(ABC):
```
Comment (AlexCuadron) [inline vcache/vcache_core/cache/eviction_policy/eviction_policy.py:17]: For all strategies, no explicit handling for empty metadata lists.

Comment (AlexCuadron) [reply vcache/vcache_core/cache/eviction_policy/eviction_policy.py:17]: Also, no handling for eviction_percentage = 0.

**Review by AlexCuadron — CHANGES_REQUESTED** [review]
Quite a big PR, some major issues inside eviction policy class.

File: vcache/vcache_core/cache/eviction_policy/eviction_policy.py (hunk 11 above)
Comment (luis-gasparschroeder) [reply vcache/vcache_core/cache/eviction_policy/eviction_policy.py:17]: Eviction percentage and watermark must in (0,1]

File: vcache/vcache_core/cache/cache.py (hunk 10 above)
Comment (luis-gasparschroeder) [reply vcache/vcache_core/cache/cache.py:34]: This is being checked for in infer_with_cache_info

File: vcache/vcache_core/cache/eviction_policy/eviction_policy.py (hunk 3 above)
Comment (luis-gasparschroeder) [reply vcache/vcache_core/cache/eviction_policy/eviction_policy.py:142]: > Also, If cache.get_all_embedding_metadata_objects() or self.select_victims() throws an exception, the lock is never released because these operations happen before the try/finally block in _evict_victims

Fixed that, thanks

File: tests/integration/test_eviction_policy.py [hunk 12]
```diff
@@ -0,0 +1,280 @@
...
+        # adds a new item to the cache.
+        self.patcher = patch.object(
+            self.vcache.vcache_policy.cache, "get_knn", return_value=[]
+        )
+        self.patcher.start()
+
+    def tearDown(self):
+        """Clean up resources."""
+        self.patcher.stop()
+        self.policy.shutdown()
```
[the remark this answers is not in the document]
Comment (luis-gasparschroeder) [reply tests/integration/test_eviction_policy.py:61]: Good point, fixed!

File: vcache/vcache_core/cache/eviction_policy/strategies/scu.py [hunk 13]
```diff
@@ -0,0 +1,136 @@
...
+        if max_n_obs == 0:
+            max_n_obs = 1
+
+        utilities: List[Tuple[int, float]] = []
+        for meta in all_metadata:
+            if meta.t_prime is None:
+                distance: float = float("inf")
+            else:
+                n_obs_norm: float = (
+                    len(meta.observations) / max_n_obs if max_n_obs > 0 else 0.0
```
[the remark this answers is not in the document]
Comment (luis-gasparschroeder) [reply vcache/vcache_core/cache/eviction_policy/strategies/scu.py:119]: No, this is the standard way to prevent div by zero in Python

https://www.geeksforgeeks.org/python/how-to-return-0-with-divide-by-zero-in-python/

**Comment by luis-gasparschroeder** [comment]
@AlexCuadron, thanks for the review! All your comments are addressed

**Review by vcache-project — APPROVED** [review]
